### PR TITLE
Add Refresh token endpoint and update biome login endpoint

### PR DESCRIPTION
**Pull request by RyanLassigBanks**
In this PR

1. Add new endpoint for requesting new authentication tokens `POST /biome/refresh`

The purpose of this endpoint is to offer a way to get new auth tokens after the a client's auth token has expired.

2. Adds refresh token database tables
3. Adds RefreshTokenStore trait
4. Adds SplinterRefrshTokenStore object
5. Updates `POST /biome/login` to return a refresh token
6. Removes `user_id` from key_management path
7. Refactors `authorize_users` to accept a validation object and not a user id

### CODE REVIEW

File: libsplinter/src/biome/rest_api/config.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ pub struct BiomeRestConfig {
     /// Duration of JWT tokens issued by this service
     access_token_duration: Duration,
     #[cfg(feature = "biome-credentials")]
-    /// Cost for encripting users password
+    /// Cost for encrypting user's password
```
Comment (peterschwarz) [inline libsplinter/src/biome/rest_api/config.rs]: Typo in commit message (also the summary is very long).

File: libsplinter/src/biome/rest_api/config.rs [hunk 2]
```diff
@@ -28,6 +29,8 @@ pub struct BiomeRestConfig {
     issuer: String,
     /// Duration of JWT tokens issued by this service
     access_token_duration: Duration,
+
+    refresh_token_duration: Duration,
```
Comment (peterschwarz) [inline libsplinter/src/biome/rest_api/config.rs]: For consistency with the other fields, please add a comment about this field.

File: libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+pub struct SplinterRefreshTokenStore {
+    connection: ConnectionPool,
+}
+
+impl RefreshTokenStore for SplinterRefreshTokenStore {
+    fn add_token(&self, user_id: &str, token: &str) -> Result<(), RefreshTokenError> {
+        let conn = self.connection.get()?;
+
+        splinter_user::table
```
Comment (peterschwarz) [inline libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs]: This should follow the operations pattern, like other biome stores.

File: splinterd/api/static/openapi.yml [hunk 4]
```diff
@@ -897,6 +906,67 @@ paths:
...
+            application/json:
+                schema:
+                  $ref: '#/components/schemas/ErrorBiome'
+        401:
+          description: Access token is invalid
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ErrorBiome'
+        403:
```
Comment (peterschwarz) [inline splinterd/api/static/openapi.yml]: Do more endpoints now return a 403 on invalid refresh tokens?

File: libsplinter/src/biome/rest_api/config.rs (hunk 1 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/config.rs]: The "typo" in the commit message was the typo being fixed in the commit. I've updated the message to make that clearer

File: splinterd/api/static/openapi.yml (hunk 4 above)
Comment (RyanLassigBanks) [reply splinterd/api/static/openapi.yml]: The refresh token is only passed to the `/tokens` endpoint, so currently that's the only endpoint in biome that returns a 403.

File: libsplinter/src/biome/rest_api/config.rs (hunk 2 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/config.rs]: updated

File: libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs (hunk 3 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs]: I also have serious reservations about implementing a trait for a struct in a different file or module than the one that it's defined in. I think that should be avoided whenever possible, and we should follow this layout whenever possible.

```
struct Foo {}
impl Foo {}
impl Bar for Foo {}
impl Bizz for Foo {}
```

Comment (peterschwarz) [reply libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs]: The operations pattern allows us to target as broad a slice of diesel-supported databases as possible at the operation level.  For example, we can query both PostgreSQL and sqlite using a single implementation, but inserts may require different implementations, depending on the table structure or insert statements. 

For example, sqlite doesn't support default column values and PostgreSQL does.  In the `UserStore`, this will require two implementations of the `UserStoreAddUserOperation` for each type of connection. 

As for the second question, I'm not sure I understand what's raising the objection.  Can you elaborate with a more concrete example?

Comment (RyanLassigBanks) [reply libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs]: I issue is with how the operations object is defined [here](https://github.com/Cargill/splinter/blob/master/libsplinter/src/biome/credentials/store/diesel/operations/mod.rs), but its trait implementations are spread across the operations module. For examle,

https://github.com/Cargill/splinter/blob/03e9bfaaba4a4070ad5060628901566ac53b57f3/libsplinter/src/biome/credentials/store/diesel/operations/fetch_credential_by_username.rs#L28

Basically, I think those implementations should be in on module along side the operations definition.

Comment (vaporos) [reply libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs]: The operations pattern is necessary to deal with how Diesel works -- it's complex, but essentially boils down to being able to pick/choose which implementations of operations work for each backend database. For example, we may have 4 of 5 operations work for both, and 1 of 5 different between postgresql and sqlite.

Comment (vaporos) [reply libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs]: re:Organization of operations - follow the currently pattern, it has been carefully considered. If we change it at some point, we will do so across all of libsplinter at once. That discussion can happen outside this PR, and needs to occur after the sqlite implementation has been finished.

File: libsplinter/src/rest_api/sessions/mod.rs [hunk 5]
```diff
@@ -78,7 +78,7 @@ where
     extra_validation(claims)
 }
 
-fn default_validation(issuer: &str) -> Validation {
+pub(crate) fn default_validation(issuer: &str) -> Validation {
```
Comment (vaporos) [inline libsplinter/src/rest_api/sessions/mod.rs]: Should this be restricted to biome?

File: libsplinter/src/biome/rest_api/actix/key_management.rs [hunk 6]
```diff
@@ -38,7 +38,7 @@ pub fn make_key_management_route(
     key_store: Arc<dyn KeyStore<Key>>,
     secret_manager: Arc<dyn SecretManager>,
 ) -> Resource {
-    Resource::build("/biome/users/{user_id}/keys")
+    Resource::build("/biome/users/keys")
```
Comment (vaporos) [inline libsplinter/src/biome/rest_api/actix/key_management.rs]: This change should really be in a separate PR, because it is going to require additional work to look at how we are doing routes to see if this is the only one we are going to change.

File: libsplinter/src/biome/datastore/diesel/models.rs [hunk 7]
```diff
@@ -91,3 +91,21 @@ impl From<SplinterUser> for UserModel {
         UserModel { id: user.id() }
     }
 }
+
```
Comment (shannynalayna) [inline libsplinter/src/biome/datastore/diesel/models.rs]: Small typo in commit message: s/ grated / created ? Also, need to add a period at the end of the message.

File: libsplinter/src/biome/rest_api/config.rs [hunk 8]
```diff
@@ -52,6 +59,7 @@ impl BiomeRestConfig {
 pub struct BiomeRestConfigBuilder {
     issuer: Option<String>,
     access_token_duration: Option<Duration>,
+    refresh_token_duration: Option<Duration>,
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/config.rs]: Should this have a feature-guard, same with other instances of the `refresh_token_duration` within this file ?

File: libsplinter/src/biome/rest_api/mod.rs [hunk 9]
```diff
@@ -125,8 +127,14 @@ impl RestResourceProvider for BiomeRestResourceManager {
                 ));
                 resources.push(make_login_route(
                     credentials_store.clone(),
+                    self.refresh_token_store.clone(),
+                    self.rest_config.clone(),
+                    Arc::new(AccessTokenIssuer::new(
+                        self.token_secret_manager.clone(),
+                        self.refresh_token_secret_manager.clone(),
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/mod.rs]: Also need to add a feature guard for this, which includes the biome-refresh-tokens

File: libsplinter/Cargo.toml [hunk 10]
```diff
@@ -90,6 +90,7 @@ experimental = [
     "biome-credentials",
     "biome-key-management",
     "biome-notifications",
+    "biome-refresh-tokens",
```
Comment (shannynalayna) [inline libsplinter/Cargo.toml]: Should this be separated into a separate commit, since this feature guard appears to have been used before this commit?

File: libsplinter/src/biome/rest_api/mod.rs [hunk 11]
```diff
@@ -115,6 +114,7 @@ impl RestResourceProvider for BiomeRestResourceManager {
 
         #[cfg(all(
             feature = "biome-credentials",
+            feature = "biome-refresh-tokens",
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/mod.rs]: Are the rest of the routes here guarded by the biome-refresh-tokens feature? It doesn't seem like they are since this includes the credentials list route and the users routes.

File: libsplinter/src/biome/rest_api/mod.rs [hunk 12]
```diff
@@ -70,10 +70,9 @@ use crate::rest_api::secrets::{AutoSecretManager, SecretManager};
 pub use config::{BiomeRestConfig, BiomeRestConfigBuilder};
 pub use error::BiomeRestResourceManagerBuilderError;
 
-#[cfg(all(feature = "biome-credentials", feature = "rest-api-actix"))]
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/mod.rs]: Shouldn't be deleted, just add the biome-refresh-tokens guard.

File: libsplinter/src/biome/rest_api/mod.rs [hunk 13]
```diff
@@ -78,6 +80,7 @@ pub use error::BiomeRestResourceManagerBuilderError;
 ))]
 use self::actix::{
     login::make_login_route,
+    register::make_register_route,
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/mod.rs]: It doesn't look like this should be added in this commit, but one of the previous commit. Also, it looks like the `make_register_route` is the only one that should be behind the `biome-refresh-tokens` feature. It's also added twice within this file.

File: libsplinter/src/biome/rest_api/actix/login.rs [hunk 14]
```diff
@@ -15,6 +15,7 @@
 use std::sync::Arc;
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/actix/login.rs]: S / a auth / an auth in commit message

File: libsplinter/src/biome/rest_api/actix/login.rs [hunk 15]
```diff
@@ -36,6 +37,7 @@ use super::super::resources::credentials::UsernamePassword;
 ///   }
 pub fn make_login_route(
     credentials_store: Arc<SplinterCredentialsStore>,
+    refresh_token_store: Arc<dyn RefreshTokenStore>,
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/actix/login.rs]: Either these lines with the refresh token store, or the login file itself, needs the 'biome-refresh-tokens` feature guard.

File: libsplinter/src/biome/rest_api/actix/mod.rs [hunk 16]
```diff
@@ -23,5 +23,11 @@ pub(super) mod key_management;
 pub(super) mod login;
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/actix/mod.rs]: Commit message: s/ a auth / an auth

File: libsplinter/src/biome/datastore/diesel/models.rs (hunk 7 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/datastore/diesel/models.rs]: updated

File: libsplinter/src/biome/rest_api/config.rs (hunk 8 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/config.rs]: Guarded by feature

File: libsplinter/Cargo.toml (hunk 10 above)
Comment (RyanLassigBanks) [reply libsplinter/Cargo.toml]: Moved commit earlier in history, so this is introduced earlier

File: libsplinter/src/biome/rest_api/actix/login.rs (hunk 14 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/actix/login.rs]: Fixed typo

File: libsplinter/src/biome/rest_api/actix/mod.rs (hunk 16 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/actix/mod.rs]: fixed type

File: libsplinter/src/biome/rest_api/mod.rs (hunk 9 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/mod.rs]: I don't think adding the feature guard here is necessary since, the entire match statement is feature guarded

File: libsplinter/src/biome/rest_api/mod.rs (hunk 11 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/mod.rs]: Re arranged commits so this gets added when the refresh token store gets added to the login route

File: libsplinter/src/biome/rest_api/mod.rs (hunk 12 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/mod.rs]: rearranged commits so this makes sense

File: libsplinter/src/biome/rest_api/mod.rs (hunk 13 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/mod.rs]: Fixed commits

File: libsplinter/src/biome/rest_api/actix/login.rs (hunk 15 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/actix/login.rs]: Guarded where it's used it Biome Rest Resource provider

File: libsplinter/src/rest_api/sessions/mod.rs (hunk 5 above)
Comment (RyanLassigBanks) [reply libsplinter/src/rest_api/sessions/mod.rs]: We would have to move them into the biome crate, and I think the sessions stuff should stay together in the rest-api module.

File: libsplinter/src/biome/rest_api/actix/key_management.rs (hunk 6 above)
Comment (RyanLassigBanks) [reply libsplinter/src/biome/rest_api/actix/key_management.rs]: Are you referring to the updates to route or the updates to `authorize_user` as a whole. Those changed were need to write the refresh token endpoint.

File: libsplinter/src/biome/rest_api/mod.rs (hunk 9 above)
Comment (shannynalayna) [reply libsplinter/src/biome/rest_api/mod.rs]: Ok, didn't realize the refresh token secret manager isn't necessarily behind the refresh-tokens feature.

File: libsplinter/src/biome/rest_api/actix/token.rs [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+
+                HttpResponse::Ok()
+                    .json(json!({ "token": token }))
+                    .into_future()
+            }))
+        })
+}
+
+#[derive(Deserialize)]
+pub struct RefreshToken {
```
Comment (shannynalayna) [inline libsplinter/src/biome/rest_api/actix/token.rs]: This should probably get moved into the resources module, under `resources/token` to follow the pattern that's currently used

File: libsplinter/src/biome/rest_api/actix/mod.rs [hunk 18]
```diff
@@ -27,6 +27,12 @@ pub(super) mod key_management;
...
 #[cfg(feature = "biome-credentials")]
 pub(super) mod register;
+#[cfg(all(
```
Comment (vaporos) [inline libsplinter/src/biome/rest_api/actix/mod.rs]: It should be fairly rare that things are guarded by multiple features; it implies that some of the items are cross-cutting concerns that can't be stated as dependencies. Here, I don't believe that is the case. This should probably only be guarded by biome-refresh-token with biome-refresh-token dependent on what it needs.

File: splinterd/Cargo.toml [hunk 19]
```diff
@@ -66,9 +66,10 @@ experimental = [
     "service-arg-validation"
 ]
 
-biome = ["splinter/biome", "database"]
-biome-credentials = ["splinter/biome-credentials", "json-web-tokens", "biome"]
-biome-key-management = ["splinter/biome-key-management", "json-web-tokens", "biome"]
+biome = ["splinter/biome",  "database"]
```
Comment (vaporos) [inline splinterd/Cargo.toml]: There is an additional space added here, looks like a typo.

File: splinterd/Cargo.toml [hunk 20]
```diff
@@ -66,9 +66,10 @@ experimental = [
...
+biome-credentials = ["splinter/biome-credentials", "biome-refresh-tokens", "json-web-tokens", "biome"]
```
Comment (vaporos) [inline splinterd/Cargo.toml]: It seems like biome-key-management and biome-credentials shouldn't depend on the biome-refresh-tokens feature, since there are no splinterd code changes here and the dependencies should be handled inside libsplinter itself.

Comment (vaporos) [reply splinterd/Cargo.toml]: Maybe same with json-web-tokens?

File: libsplinter/src/biome/rest_api/actix/key_management.rs (hunk 6 above)
Comment (vaporos) [reply libsplinter/src/biome/rest_api/actix/key_management.rs]: There are two things that could have been separate from this PR: a) the path change itself, which is a breaking change and completely unrelated to this PR; b) the fix which ignores {user_id} from the URL in favor of using the trusted user id.

I'm fine with it remaining in this PR at this point I guess.

File: libsplinter/src/biome/refresh_tokens/store/diesel/models.rs [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::schema::refresh_tokens;
+
+#[cfg(feature = "biome-refresh-tokens")]
```
Comment (vaporos) [inline libsplinter/src/biome/refresh_tokens/store/diesel/models.rs]: This and the one below are unnecessary as it is already checked withing the biome:: module.

Comment (RyanLassigBanks) [reply libsplinter/src/biome/refresh_tokens/store/diesel/models.rs]: Updated
